ops/model.py

```diff
@@ -887,6 +887,15 @@ def __iter__(self):
         return iter(self._data)
 
     def __getitem__(self, key: 'UnitOrApplication'):
+        if key is None and self.relation.app is None:
+            # NOTE: if juju gets fixed to set JUJU_REMOTE_APP for relation-broken events, then that
+            # should fix the only case in which we expect key to be None - potentially removing the
+            # need for this error in future ops versions (i.e. if relation.app is guaranteed to not
+            # be None. See https://bugs.launchpad.net/juju/+bug/1960934.
+            raise KeyError(
```

Comment:
Raising a KeyError rather than an ops specific error here is interesting. Does it make more sense to raise an AttributeError, since folks will generally be accessing this via dot notation?

We could also implement some sort of NotAvailableInThisHook exception, for when a charm author is in a hook that doesn't contain a given piece of information.

Regardless, the error message is good. I am +1 with the KeyError, if you think that is the best error to use.